### PR TITLE
[ConstraintSystem] Detect and diagnose extraneous return(s) in result build body

### PR DESCRIPTION
**Pull request by xedin**
This is (hopefully) the last source of "failed to produce a diagnostic" related to result builder
where having a `return` would just assume that body is a regular closure just to fail a bit later.

Instead of doing that (in diagnostic mode) let's mark body with explicit returns as invalid and
carry on solving to diagnose it if it was possible to reach a solution, this would also mean
that situations where nested result builders have `return` statements are going to be correctly
diagnosed.

Resolves: rdar://problem/66709164

### CODE REVIEW

**Comment by xedin** [comment]
@swift-ci please smoke test

File: test/Constraints/result_builder_diags.swift [hunk 1]
```diff
@@ -286,6 +286,21 @@ struct MyTuplifiedStruct {
   }
 }
 
+func test_invalid_return_type_in_body() {
+  tuplify(true) { _ -> (Void, Int) in
+    tuplify(false) { condition in
+      if condition {
+        return 42 // expected-error {{application of result builder 'TupleBuilder' disabled by explicit 'return' statement}}
```
Comment (hborla) [inline test/Constraints/result_builder_diags.swift:293]: Should the error message say something different if there isn't an option to "disable" the result builder? Something like `result builder [closure|function] cannot have explicit return statement`?

Comment (xedin) [reply test/Constraints/result_builder_diags.swift:293]: Sure, I think we can use `cannot use explicit return statement in the body of result builder %0`, WDYT?

Comment (hborla) [reply test/Constraints/result_builder_diags.swift:293]: I like that a lot more than my suggestion, thanks!

**Comment by xedin** [comment]
@swift-ci please smoke test

**Review by hborla — APPROVED** [review]

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by xedin** [comment]
@swift-ci please smoke test Windows platform

**Comment by xedin** [comment]
@swift-ci please smoke test Windows platform

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by xedin** [comment]
@swift-ci please smoke test Linux platform

**Comment by xedin** [comment]
@swift-ci please smoke test Linux platform

**Comment by xedin** [comment]
@swift-ci please smoke test Linux platform

**Pull request merged**